### PR TITLE
Add check on find_best_match_for_name and improve help message for undefined macro

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
I'm wondering if instead of a second help message, a note would be better. I let it up to reviewers.
#31660

### CODE REVIEW

File: src/test/compile-fail/macro_undefined.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+mod m {
+    #[macro_export]
+    macro_rules! kl {
+        () => ()
+    }
+}
+
+fn main() {
+    k!(); //~ ERROR macro undefined: 'k!'
+          //~^ HELP did you mean `kl!`?
```
Comment (sfackler) [inline src/test/compile-fail/macro_undefined.rs]: I'm assuming compiletest can't check help messages?

Comment (GuillaumeGomez) [inline src/test/compile-fail/macro_undefined.rs]: I added it afterward and I'm still waiting an opinion on it. But if it's good as it is right now, then yes.

**Comment by sfackler** [comment]
It seems a bit weird that the message would suggest a better name, but if you switch to that one you'll get another error since it wasn't `macro_use`d. Do we track enough information to add in the "don't forget to add #[macro_use]" message only when necessary?

**Comment by GuillaumeGomez** [comment]
In #31660, it's solved with `#[macro_use]`. But I'm not sure we could check this so easily...

**Comment by GuillaumeGomez** [comment]
Actually, if the suggested name is the same, we can assume that it's this kind of error.

**Comment by GuillaumeGomez** [comment]
@sfackler: It seems way better like this. What do you think?

**Comment by sfackler** [comment]
Yep, this looks good to me.

@bors r+

**Pull request merged**